lib/Dialect/Secret/IR/SecretOps.cpp

```diff
@@ -368,6 +372,193 @@ GenericOp GenericOp::removeYieldedValues(ArrayRef<int> yieldedIndicesToRemove,
   return cloneWithNewTypes(*this, newResultTypes, rewriter);
 }
 
+GenericOp GenericOp::extractOpBeforeGeneric(Operation *opToExtract,
+                                            PatternRewriter &rewriter) {
+  assert(opToExtract->getParentOp() == *this);
+
+  // Result types are secret versions of the results of the op, since the
+  // secret will yield all of this op's results immediately.
+  SmallVector<Type> newResultTypes;
+  newResultTypes.reserve(opToExtract->getNumResults());
+  for (Type ty : opToExtract->getResultTypes()) {
+    newResultTypes.push_back(SecretType::get(ty));
+  }
+
+  auto newGeneric = rewriter.create<GenericOp>(
+      getLoc(), getInputs(), newResultTypes,
+      [&](OpBuilder &b, Location loc, ValueRange blockArguments) {
+        IRMapping mp;
+        for (BlockArgument blockArg : getBody()->getArguments()) {
+          mp.map(blockArg, blockArguments[blockArg.getArgNumber()]);
+        }
+        auto *newOp = b.clone(*opToExtract, mp);
+        b.create<YieldOp>(loc, newOp->getResults());
+      });
+
+  // Once the op is split off into a new generic op, we need to add new
+  // operands to the old generic op, add new corresponding block arguments, and
+  // replace all uses of the opToExtract's results with the created block
+  // arguments.
+  SmallVector<Value> oldGenericNewBlockArgs;
+  rewriter.modifyOpInPlace(*this, [&]() {
+    getInputsMutable().append(newGeneric.getResults());
+    for (auto ty : opToExtract->getResultTypes()) {
+      BlockArgument arg = getBody()->addArgument(ty, opToExtract->getLoc());
+      oldGenericNewBlockArgs.push_back(arg);
+    }
+  });
+  rewriter.replaceOp(opToExtract, oldGenericNewBlockArgs);
+
+  return newGeneric;
+}
+
+// When replacing a generic op with a new one, and given an op in the original
+// generic op, find the corresponding op in the new generic op.
+//
+// Note, this is brittle and depends on the two generic ops having identical
```

Comment:
nit: just in case this ends up being used in a bad place, is it possible to debug assert that the ops have the same number of operations? i believe the usual only difference in its use is the # of yields

Reply:
Done